### ISSUE TITLE
Add Propagation interface

### ISSUE THREAD

**Issue by asanluis** [openzipkin/zipkin-js#496]
Hi guys, i was thinking to add the Propagation interface like (https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/propagation/Propagation.java), to support other propagation headers like f.e x-amzn-trace-id or b3 simple 😄.

**Comment by jcchavezs** [comment]
That is actually something great. This library does not have interfaces for propagation so it would be great to start one. Notice that currently, all the HTTP server instrumentation includes a function called `readHeader` in the `recordRequest` whose aim is to extract the headers from the values from the headers. I guess maybe that was because of the lack of standard headers in node? (just guessing here).

IMHO the easiest is to deprecate that `readHeader` and delegate the responsibility of such thing into the tracer ([like zipkin-go does](https://github.com/openzipkin/zipkin-go/blob/master/tracer.go#L156)) when `readHeader` is not passed? Ideally Tracer and Propagation should be separated like in brave or zipkin-php but this alternative does not feel too bad. If including extract/inject in the tracer is not acceptable, then probably we can inject the propagation object into httpInstrumentation (still we will have an abstraction to propagation but not inside the tracer).

Thoughts @adriancole @anuraaga

**Comment by asanluis** [comment]
Yeah, i'm going to try 😅😁!!!! When i have something i'll open a PR and wait the feedback. Thanks as always!!!

**Comment by anuraaga** [comment]
Interop is the name of the game, very happy to hear about adding support for more propagations.

Not familiar enough with the js codebase but b) feels a bit nicer if it's practical. But don't feel too strongly either way.

**Comment by jcchavezs** [comment]
@asanluis do you need any further help on this one?

**Comment by asanluis** [comment]
@jcchavezs, not enough time 😞, just waiting to have a little of free time to work in this!!😉

**Comment by asanluis** [comment]
@jcchavezs Hi, I've not forgotten 😥 but now i have a little of free time, and i want to ask, if you guys like this approach
`  namespace propagation {
    interface Propagation {
      extractor<T>(readHeader: <T> (header: string) => option.IOption<T>): TraceId;
      injector(req: { headers?: any }, traceId: TraceId): ZipkinHeader;
    }
    class B3Propagation implements Propagation {
      extractor<T>(readHeader: <T> (header: string) => option.IOption<T>): TraceId;
      injector(req: { headers?: any }, traceId: TraceId): ZipkinHeader;
    }
  }`

the extractor would be the actual _createIdFromHeaders of the HttpServer class, and the injector would be the actual Request.addZipkinHeaders

**Comment by asanluis** [comment]
Hi guys, i finally made a different approach, i add the propagation into the tracer 😢 , not the best implementation, but i think is the best way to easily add the baggage feature. i'll wait your feedback guys, Thanks for all!! [https://github.com/openzipkin/zipkin-js/compare/master...asanluis:496-Add_propagation_interface?expand=1](changes branch-master)

**Comment by jcchavezs** [comment]
Awesome. I will check it out.

ons. 15. jul. 2020, 21:15 skrev asanluis <notifications@github.com>:

**Comment by jcchavezs** [comment]
Hi @asanluis thanks for bringing it up. I think the approach is correct, I'd encourage you to open a PR with the changes so I can give you some feedback!

**Comment by asanluis** [comment]
I finally made the [PR](https://github.com/openzipkin/zipkin-js/pull/510), i have to work in the grpc client 😞

**Comment by asanluis** [comment]
@jcchavezs have you had time to review it?

**Comment by jcchavezs** [comment]
Hi @asanluis I am sorry I did not come back to you earlier. Crazy days and a job switch in the middle. I am looking at it today and in general will have more time to spend on this. Stay tuned.

**Comment by asanluis** [comment]
Hi @jcchavezs thanks for the feedback 😄, I like the Java approach of Getter and Setter in the injectors/extractor methods i added [here](https://github.com/openzipkin/zipkin-js/pull/510/commits/5349ef71878e1ba63f3b46903e21572603ac15d5), now im working in the implementation
